File: jetbrains-plugin/src/main/kotlin/com/asakii/plugin/settings/CodexSettings.kt
package com.asakii.plugin.settings

import com.intellij.openapi.components.*
import com.intellij.openapi.project.Project
import com.intellij.util.xmlb.XmlSerializerUtil

/**
 * Codex 设置服务
 *
 * 使用 IntelliJ Platform 的持久化机制保存 Codex 配置
 */
@State(
    name = "CodexSettings",
    storages = [Storage("codex-settings.xml")]
)
@Service(Service.Level.PROJECT)
class CodexSettings : PersistentStateComponent<CodexSettings> {

    /**
     * Codex 二进制文件路径
     */
    var binaryPath: String = ""

    /**
     * 模型提供者
     */
    var modelProvider: String = CodexConfigurable.ModelProvider.OPENAI.name

    /**
     * 默认沙箱模式
     */
    var sandboxMode: String = CodexConfigurable.SandboxMode.WORKSPACE_WRITE.name

    /**
     * Codex 是否已启用
     */
    var enabled: Boolean = false

    /**
     * 最后一次连接测试的结果
     */
    var lastTestResult: String = ""

    override fun getState(): CodexSettings = this

    override fun loadState(state: CodexSettings) {
        XmlSerializerUtil.copyBean(state, this)
    }

    companion object {
        /**
         * 获取项目级别的 Codex 设置实例
         */
        fun getInstance(project: Project): CodexSettings {
            return project.service<CodexSettings>()
        }
    }

    /**
     * 获取模型提供者枚举
     */
    fun getModelProviderEnum(): CodexConfigurable.ModelProvider {
        return try {
            CodexConfigurable.ModelProvider.valueOf(modelProvider)
        } catch (e: IllegalArgumentException) {
            CodexConfigurable.ModelProvider.OPENAI
        }
    }

    /**
     * 设置模型提供者
     */
    fun setModelProvider(provider: CodexConfigurable.ModelProvider) {
        modelProvider = provider.name
    }

    /**
     * 获取沙箱模式枚举
     */
    fun getSandboxModeEnum(): CodexConfigurable.SandboxMode {
        return try {
            CodexConfigurable.SandboxMode.valueOf(sandboxMode)
        } catch (e: IllegalArgumentException) {
            CodexConfigurable.SandboxMode.WORKSPACE_WRITE
        }
    }

    /**
     * 设置沙箱模式
     */
    fun setSandboxMode(mode: CodexConfigurable.SandboxMode) {
        sandboxMode = mode.name
    }

    /**
     * 检查 Codex 配置是否有效
     */
    fun isValid(): Boolean {
        if (binaryPath.isEmpty()) {
            return false
        }

        val file = java.io.File(binaryPath)
        return file.exists() && file.canExecute()
    }

    /**
     * 获取配置摘要信息
     */
    fun getSummary(): String {
        return buildString {
            append("Binary: $binaryPath\n")
            append("Provider: ${getModelProviderEnum().displayName}\n")
            append("Sandbox: ${getSandboxModeEnum().displayName}\n")
            append("Enabled: $enabled\n")
            if (lastTestResult.isNotEmpty()) {
                append("Last Test: $lastTestResult")
            }
        }
    }
}

File: vscode-extension/src/ide/settings/codexSettings.ts
/**
 * Codex Settings Service
 * 
 * Manages Codex configuration using VS Code workspace settings
 * Translated from jetbrains-plugin/src/main/kotlin/com/asakii/plugin/settings/CodexSettings.kt
 */

import * as vscode from 'vscode';
import * as fs from 'fs';
import * as path from 'path';

/**
 * Model Provider enum
 */
export enum ModelProvider {
    OPENAI = 'OPENAI',
    AZURE = 'AZURE',
    ANTHROPIC = 'ANTHROPIC'
}

export const ModelProviderDisplayName: Record<ModelProvider, string> = {
    [ModelProvider.OPENAI]: 'OpenAI',
    [ModelProvider.AZURE]: 'Azure OpenAI',
    [ModelProvider.ANTHROPIC]: 'Anthropic'
};

/**
 * Sandbox Mode enum
 */
export enum SandboxMode {
    WORKSPACE_WRITE = 'WORKSPACE_WRITE',
    WORKSPACE_READ = 'WORKSPACE_READ',
    NONE = 'NONE'
}

export const SandboxModeDisplayName: Record<SandboxMode, string> = {
    [SandboxMode.WORKSPACE_WRITE]: 'Workspace Write',
    [SandboxMode.WORKSPACE_READ]: 'Workspace Read Only',
    [SandboxMode.NONE]: 'No Sandbox'
};

/**
 * Codex Settings interface
 */
export interface CodexSettingsState {
    binaryPath: string;
    modelProvider: string;
    sandboxMode: string;
    enabled: boolean;
    lastTestResult: string;
}

const DEFAULT_SETTINGS: CodexSettingsState = {
    binaryPath: '',
    modelProvider: ModelProvider.OPENAI,
    sandboxMode: SandboxMode.WORKSPACE_WRITE,
    enabled: false,
    lastTestResult: ''
};

const CONFIG_KEY = 'claudeCodePlus.codex';

/**
 * Codex Settings Service
 * 
 * Uses VS Code workspace configuration for persistence
 */
export class CodexSettings {
    private static instance: CodexSettings | undefined;
    private changeListeners: Set<(settings: CodexSettings) => void> = new Set();

    private constructor() {}

    static getInstance(): CodexSettings {
        if (!this.instance) {
            this.instance = new CodexSettings();
        }
        return this.instance;
    }

    /**
     * Get current settings state
     */
    getState(): CodexSettingsState {
        const config = vscode.workspace.getConfiguration(CONFIG_KEY);
        return {
            binaryPath: config.get('binaryPath', DEFAULT_SETTINGS.binaryPath),
            modelProvider: config.get('modelProvider', DEFAULT_SETTINGS.modelProvider),
            sandboxMode: config.get('sandboxMode', DEFAULT_SETTINGS.sandboxMode),
            enabled: config.get('enabled', DEFAULT_SETTINGS.enabled),
            lastTestResult: config.get('lastTestResult', DEFAULT_SETTINGS.lastTestResult)
        };
    }

    /**
     * Get binary path
     */
    get binaryPath(): string {
        return this.getState().binaryPath;
    }

    /**
     * Set binary path
     */
    async setBinaryPath(value: string): Promise<void> {
        await this.updateSetting('binaryPath', value);
    }

    /**
     * Get model provider enum
     */
    getModelProviderEnum(): ModelProvider {
        const value = this.getState().modelProvider;
        return ModelProvider[value as keyof typeof ModelProvider] ?? ModelProvider.OPENAI;
    }

    /**
     * Set model provider
     */
    async setModelProvider(provider: ModelProvider): Promise<void> {
        await this.updateSetting('modelProvider', provider);
    }

    /**
     * Get sandbox mode enum
     */
    getSandboxModeEnum(): SandboxMode {
        const value = this.getState().sandboxMode;
        return SandboxMode[value as keyof typeof SandboxMode] ?? SandboxMode.WORKSPACE_WRITE;
    }

    /**
     * Set sandbox mode
     */
    async setSandboxMode(mode: SandboxMode): Promise<void> {
        await this.updateSetting('sandboxMode', mode);
    }

    /**
     * Get enabled status
     */
    get enabled(): boolean {
        return this.getState().enabled;
    }

    /**
     * Set enabled status
     */
    async setEnabled(value: boolean): Promise<void> {
        await this.updateSetting('enabled', value);
    }

    /**
     * Get last test result
     */
    get lastTestResult(): string {
        return this.getState().lastTestResult;
    }

    /**
     * Set last test result
     */
    async setLastTestResult(value: string): Promise<void> {
        await this.updateSetting('lastTestResult', value);
    }

    /**
     * Check if Codex configuration is valid
     */
    isValid(): boolean {
        const binaryPath = this.binaryPath;
        if (!binaryPath) {
            return false;
        }

        try {
            const stats = fs.statSync(binaryPath);
            return stats.isFile();
        } catch {
            return false;
        }
    }

    /**
     * Get configuration summary
     */
    getSummary(): string {
        const state = this.getState();
        const lines = [
            `Binary: ${state.binaryPath || '(not set)'}`,
            `Provider: ${ModelProviderDisplayName[this.getModelProviderEnum()]}`,
            `Sandbox: ${SandboxModeDisplayName[this.getSandboxModeEnum()]}`,
            `Enabled: ${state.enabled}`
        ];
        if (state.lastTestResult) {
            lines.push(`Last Test: ${state.lastTestResult}`);
        }
        return lines.join('\n');
    }

    /**
     * Add change listener
     */
    addChangeListener(listener: (settings: CodexSettings) => void): void {
        this.changeListeners.add(listener);
    }

    /**
     * Remove change listener
     */
    removeChangeListener(listener: (settings: CodexSettings) => void): void {
        this.changeListeners.delete(listener);
    }

    /**
     * Update a setting and notify listeners
     */
    private async updateSetting<K extends keyof CodexSettingsState>(
        key: K,
        value: CodexSettingsState[K]
    ): Promise<void> {
        const config = vscode.workspace.getConfiguration(CONFIG_KEY);
        await config.update(key, value, vscode.ConfigurationTarget.Workspace);
        this.notifyChange();
    }

    /**
     * Notify all listeners of change
     */
    private notifyChange(): void {
        this.changeListeners.forEach(listener => listener(this));
    }
}

// Export singleton instance
export const codexSettings = CodexSettings.getInstance();
